engine/crates/engine-v2/src/sources/graphql/query.rs

```diff
@@ -127,65 +166,87 @@ impl QueryBuilder {
         // selection set.
         if buffer.len() == n {
-            buffer.indent_write("__typename\n")?;
+            indent_write!(buffer, "__typename\n")?;
         }
         buffer.indent -= 1;
-        buffer.indent_write("}\n")?;
-        Ok(())
+        indent_write!(buffer, "}}\n")
     }
 
-    fn write_fragment_spread(&mut self, buffer: &mut Buffer, spread: PlanFragmentSpread<'_>) -> Result<(), Error> {
+    fn write_fragment_spread(
+        &mut self,
+        buffer: &mut Buffer,
+        parent_ty: SelectionSetTypeWalker<'_>,
+        spread: PlanFragmentSpread<'_>,
+    ) -> Result<(), Error> {
+        // We're writing a named fragment directly into the query as an inline one. The query might
+        // end up being a bit larger, but trying to keep fragments is not that efficient as their
+        // fields might have been planned differently at different locations. The other concern are
+        // named fragment only directives, but that's something we'll see if we ever need to.
         let fragment = spread.fragment();
-        // Nothing to deal with fragment cycles here, they should have been detected way earlier
-        // during query validation.
-        let mut fragment_buffer = Buffer::default();
-        // the actual name is computed afterwards as attribution of the fragment fields will depend
-        // on its spread location, so it isn't necessarily the same. Once we have tests for
-        // directives we could simplify that as there is not need to keep named fragment except for
-        // their directives that the upstream server may understand.
-        fragment_buffer.write_str(&format!("on {}", fragment.type_condition_name()))?;
-        self.write_selection_set(&mut fragment_buffer, spread.selection_set())?;
-        let name = self.fragment_content_to_name.entry(fragment_buffer).or_insert_with(|| {
-            let id = self
-                .fragment_name_to_last_id
-                .entry(fragment.name().to_string())
-                .and_modify(|id| *id += 1)
-                .or_default();
-            format!("{}_{}", fragment.name(), id)
-        });
-        buffer.indent_write(&format!("...{name}\n"))?;
+        let ty = fragment.ty();
+        if parent_ty != ty {
+            indent_write!(buffer, "... on {}", ty.name())?;
+            self.write_selection_set(buffer, spread.selection_set())?;
+        } else {
+            // We don't create a nested selection set if the type condition is equivalent to the parent
+            // type
+            for selection in spread.selection_set() {
+                match selection {
+                    PlanSelection::Field(field) => self.write_field(buffer, field)?,
+                    PlanSelection::FragmentSpread(spread) => self.write_fragment_spread(buffer, ty, spread)?,
+                    PlanSelection::InlineFragment(fragment) => self.write_inline_fragment(buffer, ty, fragment)?,
+                }
+            }
```

Comment:
What's the thinking behind this branch?  I get that we don't need the type condition, but it'd be simpler to just send it anyway and I can't see much downside.

Also same question about directives here (which would be one reason to make this an inline fragment rather than flattening)